lib/protein/adapters/amqp_adapter/server.ex

```diff
@@ -12,22 +12,37 @@ defmodule Protein.AMQPAdapter.Server do
   end
 
   def init(opts) do
+    Process.flag(:trap_exit, true)
     chan = connect(opts)
-    {:ok, {chan, opts}}
+    {:ok, {chan, opts, 0}}
+  end
+
+  def terminate(_reason, {_chan, _opts, running_processes_count}) do
+    wait_for_all_processes(running_processes_count)
+  end
+
+  def wait_for_all_processes(0), do: :ok
+
+  def wait_for_all_processes(running_processes_count) do
+    receive do
```

Comment:
What can we expect to happen if some process will get stuck forever and this recursive loop will never finish?

Reply:
I was quitting in 5 seconds anyway during my tests.

Reply:
I've included explicit shutdown after 15 second for this supervisor's child